content/blog/2018/appswitch/appswitch.md

```diff
@@ -0,0 +1,81 @@
+
+Why are we talking about a traditional application in the modern cloud-native context?  It turns out that there are significant gains to be had by enabling them to run on Kubernetes and Istio platforms.  Essentially it’s a part of the modernization journey that allows running traditional apps alongside green-field apps on the same modern platform to facilitate interoperation between the two.  As such, running WebSphere ND on Kubernetes requires a whole lot of additional cushion from AppSwitch in the way of the network environment it expects.  For the purpose of this blog however, I'll focus on the dependency ordering requirement.
+
+Simply deploying `dmgr` and node instances as pods on a Kubernetes cluster does not work.  `dmgr` and the node instances happen to have a lengthy initialization process that can take several minutes.  If they are all co-scheduled, the application typically ends up in a funny state.  When a node instance comes up and finds that `dmgr` is missing, it would take an alternate startup path.  Instead, if it had exited immediately, Kubernetes crash-loop would have taken over and perhaps the application would have come up.  But even in that case, it turns out a clean startup is not guaranteed.
+
+One `dmgr` along with its node instances is a basic deployment configuration for WebSphere ND.  Applications like IBM Business Process Manager that are built on top of WebSphere ND running in production environments include several other services.  In those configurations, there could be multilevel interdependencies.  Depending on the applications hosted by the node instances, there may be an ordering requirement among them as well.  With long service initialization times and crash-loop restarts, there is little chance for the application to start in any reasonable length of time.
+
+### Sidecar Dependency in Istio
+
+As such, Istio itself is affected by a version of the dependency ordering problem.  Since connections into and out of a service running under Istio are redirected through its sidecar proxy, an implicit dependency is created between the application service and its sidecar.  Unless, the sidecar is fully operational, all requests from and to the service would get dropped.
+
+So how do we go about addressing these issues?  One way is to defer it to the applications and say that they are supposed to be "well behaved" and implement appropriate logic to make themselves immune to startup order.  However, many applications (especially traditional ones) either timeout or deadlock if misordered.  Even for new applications, implementing one off logic for each service is substantial additional burden that is best avoided.  Service mesh should provide a way around these problems.  After all, factoring out common patterns into an underlying framework is really the point of service mesh.
```

Comment:
Doesn't the mesh's normal fault recovery logic help this situation a lot? A service will come up and try to contact its dependencies. If these aren't available, the workload's sidecar can automatically retry until things come up.

Reply:
Actually yes, but it wouldn't have worked for the application at hand since it was UDP based.